### PR TITLE
[WIP] Generic instances

### PR DESCRIPTION
**Pull request by fosskers**
To enable use of [generic-lens](https://hackage.haskell.org/package/generic-lens-1.0.0.1) over a parsed bash AST.

### CODE REVIEW

**Comment by fosskers** [comment]
CI fails for GHC 7.8 and 7.6. Even [Eta](https://eta-lang.org/) is basically GHC 7.10.x, so is there a need to support that far back?

File: src/Language/Bash/Expand.hs [hunk 1]
```diff
@@ -180,7 +181,7 @@ data TildePrefix
     | PWD               -- ^ @~+/foo@
     | OldPWD            -- ^ @~-/foo@
     | Dirs Int          -- ^ @~N@, @~+N@, @~-N@
-    deriving (Eq, Read, Show)
+    deriving (Eq, Read, Show, Generic)
```
Comment (knrafto) [inline src/Language/Bash/Expand.hs]: maybe not this one? This doesn't have `Data` or `Typeable` either.

**Comment by fosskers** [comment]
Rebased - I'm in the middle of a refactor in Aura, so once things compile I'll confirm whether removing `Generic` on `TildePrefix` affects anything.

**Comment by fosskers** [comment]
The CI failure:
```
Tests
  Properties
    brace expansion:                               FAIL
input:
\},\},\},e,rj\ {,\}u\}\{sr\{\,tq{s\{,\{\ ,}o,}{}o\{}w}v,}\{,\}n,,\}v}}jqg,\}
bash output:
},},},e,rj {}o{}w}v,}{,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tqs{o{}o{}w}v,}{,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tq{ o{}o{}w}v,}{,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tqo{}o{}w}v,}{,}n,,}v}}jqg,} },},},e,rj {}o{}w}v,}{,}n,,}v}}jqg,}
test output:
},},},e,rj }o{}w}v{,}n,,}v}}jqg,} },},},e,rj {,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tqs{o}o{}w}v{,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tqs{o{,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tq{ o}o{}w}v{,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tq{ o{,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tqo}o{}w}v{,}n,,}v}}jqg,} },},},e,rj }u}{sr{,tqo{,}n,,}v}}jqg,} },},},e,rj }o{}w}v{,}n,,}v}}jqg,} },},},e,rj {,}n,,}v}}jqg,}
FAIL (0.13s)
      *** Failed! Assertion failed (after 78 tests):
      "\\},\\},\\},e,rj\\ {,\\}u\\}\\{sr\\{\\,tq{s\\{,\\{\\ ,}o,}{}o\\{}w}v,}\\{,\\}n,,\\}v}}jqg,\\}"
      Use --quickcheck-replay=472318 to reproduce.
```

**Comment by fosskers** [comment]
Okay, seems to work alright still! Merge away :)

**Pull request merged**